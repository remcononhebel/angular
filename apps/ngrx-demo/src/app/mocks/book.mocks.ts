export const mockBooks = [
  {
    id: 'mock id 1',
    volumeInfo: { authors: ['mock authors 1'], title: 'mock title 1' },
  },
  {
    id: 'mock id 2',
    volumeInfo: { authors: ['mock authors 2'], title: 'mock title 2' },
  },
];
