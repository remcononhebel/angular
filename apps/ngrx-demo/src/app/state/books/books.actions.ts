import { createActionGroup, emptyProps, props } from '@ngrx/store';
import { Book } from '../../model/book.model';

export const BooksPageActions = createActionGroup({
  source: 'Books Page',
  events: {
    'Load Book List': emptyProps(),
  },
});

export const BooksApiActions = createActionGroup({
  source: 'Books API',
  events: {
    'Load Book List Successful': props<{ books: Book[] }>(),
    'Load Book List Failed': props<{ error: string }>(),
  },
});
