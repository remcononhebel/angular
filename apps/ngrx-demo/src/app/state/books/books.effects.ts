import { Injectable } from "@angular/core";
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { catchError, exhaustMap, map, of } from "rxjs";
import { BooksApiService } from "../../service/books-api.service";

@Injectable()
export class BooksEffects {
    loadBooks$ = createEffect(() => this.actions$.pipe(
        ofType('[Books Page] Load Book List'),
        exhaustMap(() => this.booksApiService.getBooks()
            .pipe(
                map(books => ({ type: '[Books API] Load Book List Successful', books })),
                catchError((error) => of({type: '[Books API] Load Book List Failed', error: 'Failed to load books: ' + error.message}))
            )
        )
    ))


    constructor(private readonly actions$: Actions, private readonly booksApiService: BooksApiService) { }
}