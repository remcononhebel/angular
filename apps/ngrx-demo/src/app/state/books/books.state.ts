import { Book } from '../../model/book.model';

export interface BooksState {
  books: Book[];
  error: string | null;
}
