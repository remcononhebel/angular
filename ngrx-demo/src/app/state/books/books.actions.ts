import { createActionGroup, props } from '@ngrx/store';
import { Book } from '../../model/book.model';

export const BooksApiActions = createActionGroup({
  source: 'Books API',
  events: {
    'Retrieved Book List': props<{ books: Book[] }>(),
  },
});
