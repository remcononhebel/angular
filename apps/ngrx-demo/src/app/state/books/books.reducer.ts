import { createReducer, on } from '@ngrx/store';
import { BooksApiActions } from './books.actions';
import { BooksState } from './books.state';

export const initialState: BooksState = {
  books: [],
};

export const booksReducer = createReducer(
  initialState,
  on(BooksApiActions.retrievedBookList, (state, { books }) => {
    return {
      ...state,
      books,
    };
  })
);
