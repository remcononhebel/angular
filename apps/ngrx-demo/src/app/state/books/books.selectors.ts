import { createFeatureSelector, createSelector } from '@ngrx/store';
import { BooksState } from './books.state';
import { Book } from '../../model/book.model';

export const selectBooksState = createFeatureSelector<BooksState>('books');

export const selectAvailableBooks = createSelector(
  selectBooksState,
  (booksState: BooksState) => {
    return booksState.books;
  }
);

export const selectNumberOfAvailableBooks = createSelector(
  selectAvailableBooks,
  (books: Book[]) => books.length
);
