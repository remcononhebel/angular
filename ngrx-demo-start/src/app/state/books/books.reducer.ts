import { createReducer } from '@ngrx/store';
import { BooksState } from './books.state';

export const initialState: BooksState = {
  books: [
    {
      id: '1',
      volumeInfo: {
        authors: ['Author 1', 'Co-author 1'],
        title: 'Book 1',
      },
    },
    {
      id: '2',
      volumeInfo: {
        authors: ['Author 2', 'Co-author 2'],
        title: 'Book 2',
      },
    },
  ],
};

export const booksReducer = createReducer(initialState);
