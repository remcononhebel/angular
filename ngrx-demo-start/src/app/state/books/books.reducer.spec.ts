import { mockBooks } from '../../mocks/book.mocks';
import { BooksApiActions } from './books.actions';
import { booksReducer } from './books.reducer';
import { BooksState } from './books.state';

describe('booksReducer', () => {
  it('should handle BooksApiActions.retrievedBookList', () => {
    const state: BooksState = {
      books: [],
    };

    const books = [...mockBooks];
    const action = BooksApiActions.retrievedBookList({ books });
    expect(booksReducer(state, action)).toMatchSnapshot();
  });
});
