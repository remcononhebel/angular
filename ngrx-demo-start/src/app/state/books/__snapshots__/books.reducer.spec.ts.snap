// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`booksReducer should handle BooksApiActions.retrievedBookList 1`] = `
{
  "books": [
    {
      "id": "mock id 1",
      "volumeInfo": {
        "authors": [
          "mock authors 1",
        ],
        "title": "mock title 1",
      },
    },
    {
      "id": "mock id 2",
      "volumeInfo": {
        "authors": [
          "mock authors 2",
        ],
        "title": "mock title 2",
      },
    },
  ],
}
`;
